File: ruby/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


require_relative "jugador"
require_relative 'gestor_estados'
require_relative 'estados_juego'
require_relative 'dado'
require_relative 'sorpresa'
require_relative 'tablero'
require_relative 'titulo_propiedad'
require_relative 'diario'
require_relative "mazo_sorpresas"
require_relative "tipo_sorpresa"
require_relative "casilla"


module Civitas
  class CivitasJuego
    
    attr_accessor :jugadores, :indice_jugador_actual
    
    
    def initialize(jugadores)
      @jugadores = Array.new
      for nombre in jugadores
        jugador = Jugador.new_con_nombre(nombre)
        @jugadores<<jugador
      end
      
      
      @gestor_estados = Gesor_estados.new
      @estado = @gestor_estados.estado_inicial
      @indice_jugador_actual = Dado.get_instance.quien_empieza(@jugadores.length)
      @mazo = Mazo_sorpresas.new
      inicializa_mazo_sorpresas
      @tablero = Tablero.new(Casilla.carcel)
      inicializa_tablero
      
    end
    
    def inicializa_tablero(mazo)
      #No se como funciona
    end
    
    def inicializa_mazo_sorpresas(tablero)
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::IRCARCEL,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::IRCASILLA,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PAGARCOBRAR,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PORCASAHOTEL,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PORJUGADOR,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::SALIRCARCEL,tablero))
    end
    
    def contabilizar_pasos_por_salida(jugador_actual)
      contador = @tablero.get_por_salida
      while contador > 0
        jugador_actual.pasa_por_salida
        contador = @tablero.get_por_salida
      end
    end
    
    def pasar_turno
      if (@indice_jugador_actual == (@jugadores.length -1))
        @indice_jugador_actual = 0
      else
        @indice_jugador_actual += 1
      end
    end
    
    def siguiente_paso_completado(operacion)
      @estado = @gestor_estados.siguiente_estado(get_jugador_actual, @estado, operacion)
    end
    
    def get_jugador_actual
      return @jugadores[@indice_jugador_actual]
    end
    
    def construir_casa(ip)
      @jugadores[@indice_jugador_actual].construir_casa(ip)
    end
    
    def construir_hotel(ip)
      @jugadores[@indice_jugador_actual].construir_hotel(ip)
    end
    
    def vender(ip)
      @jugadores[@indice_jugador_actual].vender(ip)
    end
    
    def hipotecar(ip)
      @jugadores[@indice_jugador_actual].hipotecar(id)
    end
    
    def cancelar_hipoteca(id)
      @jugadores[@indice_jugador_actual].cancelar_hipoteca(id)
    end
    
    def salir_carcel_pagando
      @jugadores[@indice_jugador_actual].salir_carcel_pagando
    end
    
    def salir_carcel_tirando
      @jugadores[@indice_jugador_actual].salir_carcel_tirando
    end
    
    def final_del_juego
      bancarrota = false
      for jugador in @jugadores
        if jugador.en_bancarrota
          bancarrota = true
        end
      end
      return bancarrota
    end
    
    def ranking
      @jugadores.sort! { |a,b| a.saldo <=> b.saldo }
    end
    
  end
end

File: java/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Random;
import java.util.Collections;
/**
 *
 * @author mota
 */
public class CivitasJuego {
    private int indiceJugadorActual;
    
    //asociaciones
    private MazoSorpresas mazoS;
    private Tablero tablero;
    private ArrayList<Jugador> jugadores;
    private GestorEstados gestorE;
    private EstadosJuego estado;
    
    //constructor
    
    CivitasJuego(ArrayList<String> nombres){
        jugadores = new ArrayList<Jugador>();
        
        for(String n:nombres){
            jugadores.add(new Jugador(n));
        }
        
        mazoS = new MazoSorpresas();
        
        gestorE = new GestorEstados();
        
        estado = gestorE.estadoInicial();
        
        indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
        
        //inicializar
        //inicializaTablero(mazoS);
        inicializarMazoSorpresa(tablero);
        
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(Casilla.getCarcel());
        
        for(int i=1; i<3; i++){
            TituloPropiedad propiedad = new TituloPropiedad("calle"+i,100*i,3,50*i,500+100*i,200);
            Casilla cas = new Casilla(propiedad);
            tablero.añadeCasilla(cas);
        }
        
        Casilla cas1;
        
        cas1 = new Casilla(mazoS,"SorpresaCasilla");
        tablero.añadeCasilla(cas1);
        tablero.añadeJuez();
        
    }
    
    private void inicializarMazoSorpresa(Tablero tablero){
        
        Random random = new Random();
        
        for(TipoSorpresa tipo:TipoSorpresa.values()){
        
            switch(tipo){
                case IRCARCEL:
                    mazoS.alMazo(new Sorpresa(tipo,tablero));
                break;
                case SALIRCARCEL:
                    mazoS.alMazo(new Sorpresa(tipo,mazoS));
                break;
                case IRCASILLA:
                    
                    mazoS.alMazo(new Sorpresa(tipo,tablero,random.nextInt(tablero.size()),tipo.toString()));
                break;
                default:
                    mazoS.alMazo(new Sorpresa(tipo,random.nextInt(400)+100,tipo.toString()));

            }
        
        }
    }
    
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida()>0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    private void pasarTurno(){
        if(indiceJugadorActual<=jugadores.size()){
            indiceJugadorActual++;
        }else{
            indiceJugadorActual=0;
        }
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorE.siguienteEstado(getJugadorActual(), estado, operacion);
    }
    
    public boolean vender(int ip){
        return jugadores.get(indiceJugadorActual).vender(ip);
    }
    
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    }
    
    /*
    public boolean cancelarHipotecta(){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca();
    }
    */
    /*
    public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
    } 
    */
    /*
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
    } 
    */
    /*
    public boolean hipotecar(int ip){
        return jugadores.get(indiceJugadorActual).hipotecar();
    } 
    */
            
    private ArrayList<Jugador> ranking(){
        Collections.sort(jugadores);
        return jugadores;
    }
    
    public Casilla getCasillaActual(){
        return tablero.getCasilla(jugadores.get(indiceJugadorActual).getNumCasillaActual());
    }
    
   public String infoJugadorTexto(){
       return jugadores.get(indiceJugadorActual).toString();
   }
   
   public boolean finalDelJuego(){
       boolean resultado = false;
       
       //recorrer el array jugadores
       for(int i=0; i<jugadores.size(); i++){
           resultado = jugadores.get(i).enBancarrota();
       }
       return resultado;
   }
}
